codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java

```diff
@@ -300,74 +291,108 @@ public static String decodeComponent(final String s, final Charset charset) {
         if (s == null) {
             return EMPTY_STRING;
         }
-        final int size = s.length();
-        boolean modified = false;
-        for (int i = 0; i < size; i++) {
-            final char c = s.charAt(i);
-            if (c == '%' || c == '+') {
-                modified = true;
+        return decodeComponent(s, 0, s.length(), charset, false);
+    }
+
+    private static String decodeComponent(String s, int from, int toExcluded, Charset charset, boolean isPath) {
+        int len = toExcluded - from;
+        if (len <= 0) {
+            return EMPTY_STRING;
+        }
+        int firstEscaped = -1;
+        for (int i = from; i < toExcluded; i++) {
+            char c = s.charAt(i);
+            if (c == '%' || c == '+' && !isPath) {
+                firstEscaped = i;
                 break;
             }
         }
-        if (!modified) {
-            return s;
+        if (firstEscaped == -1) {
+            return s.substring(from, toExcluded);
         }
-        final byte[] buf = new byte[size];
-        int pos = 0;  // position in `buf'.
-        for (int i = 0; i < size; i++) {
+
+        CharsetDecoder decoder = CharsetUtil.decoder(charset);
+        int decodedCapacity = len / 3;
```

Comment:
can you add a comment explaining `len / 3`? something to the effect of each encoded byte takes 3 characters.